File: snapshot/2009/SDK/VSTA Samples/GetTimeElapsedSinceLastSave/VB.NET/ThisApplication.vb

<CLSCompliant(False)> _
Partial Class ThisApplication

    Dim m_dicLastSaved As System.Collections.Generic.Dictionary(Of Autodesk.Revit.Document, DateTime) = New System.Collections.Generic.Dictionary(Of Autodesk.Revit.Document, DateTime)

    Private Sub ThisApplication_Startup(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Startup
        InitializeDicLastSaved()
        AddHandler Me.OnDocumentNewed, AddressOf ThisApplication_OnDocumentNewed
        AddHandler Me.OnDocumentClosed, AddressOf ThisApplication_OnDocumentClosed
        AddHandler Me.OnDocumentOpened, AddressOf ThisApplication_OnDocumentOpened
        AddHandler Me.OnDocumentSaved, AddressOf ThisApplication_OnDocumentSaved
    End Sub

    Private Sub ThisApplication_Shutdown(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Shutdown
        RemoveHandler Me.OnDocumentNewed, AddressOf ThisApplication_OnDocumentNewed
        RemoveHandler Me.OnDocumentClosed, AddressOf ThisApplication_OnDocumentClosed
        RemoveHandler Me.OnDocumentOpened, AddressOf ThisApplication_OnDocumentOpened
        RemoveHandler Me.OnDocumentSaved, AddressOf ThisApplication_OnDocumentSaved
    End Sub

    Sub ThisApplication_OnDocumentNewed(ByVal document As Autodesk.Revit.Document)
        m_dicLastSaved.Add(document, DateTime.MaxValue)
    End Sub 'ThisApplication_OnDocumentNewed


    Sub ThisApplication_OnDocumentClosed(ByVal document As Autodesk.Revit.Document)
        m_dicLastSaved.Remove(document)
    End Sub 'ThisApplication_OnDocumentClosed


    Sub ThisApplication_OnDocumentOpened(ByVal document As Autodesk.Revit.Document)
        m_dicLastSaved.Add(document, DateTime.MaxValue)
    End Sub 'ThisApplication_OnDocumentOpened


    Sub ThisApplication_OnDocumentSaved(ByVal document As Autodesk.Revit.Document)
        Me.m_dicLastSaved(document) = DateTime.Now
    End Sub 'ThisApplication_OnDocumentSaved

    Private Sub InitializeDicLastSaved()
        Dim document As Autodesk.Revit.Document
        For Each document In Me.Documents
            m_dicLastSaved.Add(document, DateTime.MaxValue)
        Next document
    End Sub 'InitializeDicLastSaved

    Private Function FormatTimeSpan(ByVal elapse As TimeSpan) As String
        Dim elapseStr As String = elapse.ToString()
        Dim lastIndexOfDot As Integer = elapseStr.LastIndexOf(".")
        Return elapseStr.Substring(0, lastIndexOfDot)
    End Function 'FormatTimeSpan

    Public Sub GetTimeElapsedSinceLastSave()
        Dim text As String = String.Format("{0,-30}{1,-30}", "Document Full Name", "Elapse Time Since Last Save(day.hh:mm:ss)") + ControlChars.Lf
        Dim pair As System.Collections.Generic.KeyValuePair(Of Autodesk.Revit.Document, DateTime)
        For Each pair In m_dicLastSaved
            Dim strElapsed As String = String.Empty
            If pair.Value = DateTime.MaxValue Then
                strElapsed = "Never"
            Else
                strElapsed = FormatTimeSpan((DateTime.Now - pair.Value))
            End If

            Dim fileName As [String] = System.IO.Path.GetFileName(pair.Key.PathName)
            If [String].IsNullOrEmpty(fileName) Then
                fileName = "*New*"
            End If

            text += String.Format("{0,-30}{1,-30}", fileName, strElapsed) + ControlChars.Lf
        Next pair

        System.Windows.Forms.MessageBox.Show([text], "Macro GetTimeElapsedSinceLastSave")

    End Sub 'GetTimeElapsedSinceLastSave

End Class

File: snapshot/2010/SDK_2/VSTA Samples/GetTimeElapsed_CSharp/Source/GetTimeElapsed_CSharp/ThisApplication.cs
using System;
using Autodesk;
using Autodesk.Revit;
using Autodesk.Revit.Elements;
using System.IO;
using System.Diagnostics;
using System.Collections.Generic;

namespace GetTimeElapsed_CSharp
{
    [System.AddIn.AddIn("ThisApplication", Version = "1.0", Publisher = "", Description = "")]
    public partial class ThisApplication
    {
        Dictionary<Document, DateTime> m_dicLastSaved = new Dictionary<Document, DateTime>();

        private void Module_Startup(object sender, EventArgs e)
        {
            InitializeDicLastSaved();
            this.OnDocumentSaved += new Autodesk.Revit.Events.DocumentSavedEventHandler(ThisApplication_OnDocumentSaved);
            this.OnDocumentOpened += new Autodesk.Revit.Events.DocumentOpenedEventHandler(ThisApplication_OnDocumentOpened);
            this.OnDocumentNewed += new Autodesk.Revit.Events.DocumentNewedEventHandler(ThisApplication_OnDocumentNewed);
            this.OnDocumentClosed += new Autodesk.Revit.Events.DocumentClosedEventHandler(ThisApplication_OnDocumentClosed);
        }

        private void Module_Shutdown(object sender, EventArgs e)
        {
            this.OnDocumentSaved -= new Autodesk.Revit.Events.DocumentSavedEventHandler(ThisApplication_OnDocumentSaved);
            this.OnDocumentOpened -= new Autodesk.Revit.Events.DocumentOpenedEventHandler(ThisApplication_OnDocumentOpened);
            this.OnDocumentNewed -= new Autodesk.Revit.Events.DocumentNewedEventHandler(ThisApplication_OnDocumentNewed);
            this.OnDocumentClosed -= new Autodesk.Revit.Events.DocumentClosedEventHandler(ThisApplication_OnDocumentClosed);
        }

        #region VSTA generated code
        private void InternalStartup()
        {
            this.Startup += new System.EventHandler(Module_Startup);
            this.Shutdown += new System.EventHandler(Module_Shutdown);
        }
        #endregion

        private void InitializeDicLastSaved()
        {
            foreach (Document document in this.Documents)
            {
                m_dicLastSaved.Add(document, DateTime.MaxValue);
            }
        }

        void ThisApplication_OnDocumentNewed(Document document)
        {
            m_dicLastSaved.Add(document, DateTime.MaxValue);
        }

        void ThisApplication_OnDocumentClosed(Document document)
        {
            m_dicLastSaved.Remove(document);
        }

        void ThisApplication_OnDocumentOpened(Document document)
        {
            m_dicLastSaved.Add(document, DateTime.MaxValue);
        }

        void ThisApplication_OnDocumentSaved(Document document)
        {
            this.m_dicLastSaved[document] = DateTime.Now;
        }

        private String FormatTimeSpan(TimeSpan elapse)
        {
            String elapseStr = elapse.ToString();
            int lastIndexOfDot = elapseStr.LastIndexOf('.');
            return elapseStr.Substring(0, lastIndexOfDot);
        }

        /// <summary>
        /// GetTimeElapsedSinceLastSave
        /// </summary>
        public void GetTimeElapsedSinceLastSave()
        {
            String text = String.Format("{0,-30}{1,-30}",
                "Document Full Name",
                "Elapse Time Since Last Save(day.hh:mm:ss)")
                + "\n";
            foreach (KeyValuePair<Document, DateTime> pair in m_dicLastSaved)
            {
                String strElapsed = String.Empty;
                if (pair.Value == DateTime.MaxValue)
                {
                    strElapsed = "Never";
                }
                else
                {
                    strElapsed = FormatTimeSpan(DateTime.Now - pair.Value);
                }

                String fileName = System.IO.Path.GetFileName(pair.Key.PathName);
                if (String.IsNullOrEmpty(fileName))
                {
                    fileName = "*New*";
                }

                text += String.Format("{0,-30}{1,-30}", fileName, strElapsed) + "\n";
            }

            System.Windows.Forms.MessageBox.Show(text, "Macro GetTimeElapsedSinceLastSave");
        }
    }
}
